### ISSUE TITLE
Climate scheduler

### ISSUE THREAD

**Issue by alex-torregrosa** [home-assistant/architecture#148]
_Triggered from: home-assistant/home-assistant/pull/21017_

It would be useful to add a scheduler to the climate component, specially for the generic_thermostat platform. You can actually do this by using automations in HA, but this way the ui would be more user-friendly.

It should enable the user to set rules triggering temperature changes at certain hours and days of the week. It could also be useful to add a sort of timer to manual temperature changes (making them last 1h for example), but that should be optional and enabled in the component configuration.

This could be done as a standalone componenent (see PR at the beggining) or integrate the functionality into the climate component. The standalone component approach would make sharing a single schedule between different climate entities easier.

**Comment by ardeus-ua** [comment]
Workday sensor usage will also be useful

**Comment by awarecan** [comment]
I would like to see a "smart" scheduler be implemented in future, something can compete with Nest's "learning thermostat". That scheduler should consider the weather forecast, the current temperature, and user usage pattern, etc.

If you want to implement a "time" scheduler component with a user friendly UI, I would suggest we implement a generalized scheduler component, not only for climate, but can call any service at a given time.

**Comment by alex-torregrosa** [comment]
The generalized scheduler component sounds good. However, how can we make it work with the different components? Multiple platforms for a schedule component? It is way different to call 'switch.turn_on' than 'climate.set_temperature'. The differences would also affect the UI.

If you think this is the way to go, I can try to refactor the existing code and turn it to a generic platform (and I think a new PR would be needed, am I correct?).

I also like the idea of the 'smart' scheduler, but It may be something more for the long term, and maybe it could be extended to other components as well. Fore the climate component, I have been thinking of applying some machine learning to the turn-on times of the 'generic_thermostat', as it overshoots the set temperature (When it turns off, the radiators are still hot, warming the house a little bit more). However, I don't have enough data collected yet to try anything.

**Comment by alex-torregrosa** [comment]
I've been thinking of another approach:

Instead of showing a component with a card, a new panel could be added, which would allow to have a single 'schedule' component with more complex UI options that unifies events for multiple platforms.

For example, you should be able to select a time and day/days/workday/holiday and then a event to fire. Simple events like turning on/off a light or changing a climate's temperature should have a predefined UI, and perhaps an editor for custom events.

The problem with this approach is that it would be like reinventing automations...

**Comment by efficiosoft** [comment]
Hi,

As the author of [Schedy](https://hass-apps.readthedocs.io/en/stable/apps/schedy/index.html), I just want to add my two cents here.

While it's certainly easy to design an UI for a simple scheduler with time slots for different workdays, maybe even weeks, months etc. I can't imagine a way to account for external conditions such as weather sensors, presence and that sort of things without creating if ... then... rules, which we already have with automations, although they become very complicated to maintain when used for scheduling.

What has to be considered as well is what happens when HA wasn't running at the time the scheduled value switched, there has then to be a mechanism to find out what to set at startup, so whatever the scheduling algorithm looks like, it has to be able to find a suitable value at every time.

I once also thought about integrating Schedy as a component into HA directly, but had no time to dive into it, hence it still requires AppDaemon. Since the rules concept is quite powerful, maybe that would also be an option?

Best regards
Robert

**Comment by alex-torregrosa** [comment]
Thanks Robert!

I think that the easiest thing to do would be to integrate into HA just the time-based rule functionality of Schedy. If someone wants to write more complex rules, Schedy or automations can be used.

**Comment by efficiosoft** [comment]
Yeah, at least that's the part which it should be relatively straightforward to create an UI for.

However, automations aren't suited well for scheduling IMHO. When you just want to execute service X at time Y, that's straightforward. But I think it'd be more appropriate to have a mapping between - let's say - a time slot and the state of an arbitrary entity to be maintained during that time slot. How to bring the entity to that state then differs from entity type to type.

At least that's the concept used in Schedy. You have actors (entities) of a specific type (thermostat, switch, fan or any other customizable type) and a schedule that maps time slots (and, if you go the advanced route via expressions, also arbitrary state conditions) to states of the specific actor type. It can then easily ensure all actors are in the state dictated by the schedule while still allowing for manual adjustments and, if desired, revert these adjustments after a given time.

All in all, the code isn't that small (~4000 lines) and, at the moment, tightly coupled with the AppDaemon API. I also wished to move away from using AppDaemon for it, but simply don't have the time to work on it.

**Comment by efficiosoft** [comment]
Just an idea for the simple time-based scheduling... Can't this be acomplished with a calendar + an automation that triggers when an event in the calendar starts? Maybe even the temperature/value/whatever could be supplied with the event.

That way no new interface for creating the schedules has to be written, since there are plenty of good UIs for managing calendars out there.

**Comment by timstanley1985** [comment]
Hi, I have implemented heating timers using an automation and some input selects and scripts to set the profile. Have been using them for 18 months now and they are very stable.  I am also pretty happy with how it looks in the UI too....  I can share the code if you are interested.

![screenshot_20190216-183426](https://user-images.githubusercontent.com/6829542/52903826-24896280-321b-11e9-9494-ad9fa6941b3d.png)

**Comment by pszafer** [comment]
@timstanley1985 can you share your code of timers?

**Comment by timstanley1985** [comment]
I have pasted the automations, input selects and input booleans below.  You will just need to create a clock.yaml file which is a list of available times (i have every 15 mins from 05:00 - 22:00) and a profiles.yaml which is a list of available profiles.  You then need to create a script with a name that matches each profile.  For example, script.heating_off that turns off the heating.  script.heating_medium which sets the heating to a medium setting etc.  I have per room control on my set up so have about 6 profiles for heating different parts of the house.

You can then copy and paste these automations, increasing the number by 1, and create any number of timers,  I have 9 currently.

```
########## Heating rules ##########
automation:
 - alias: Heat Profile 1
   initial_state: 'on'
   trigger:
    - platform: template
      value_template: '{%if  states.input_select.heating_timer_1_time.state == states.sensor.time.state %} true {% endif %}'
   condition:
    - condition: state
      entity_id: input_boolean.heating_timer_1_switch
      state: 'on'
    - condition: state
      entity_id: input_boolean.summer_mode
      state: 'off'
    - condition: or
      conditions:
       - condition: state
         entity_id: input_boolean.heating_holiday_mode
         state: 'off'
       - condition: state
         entity_id: input_select.heating_timer_1_profile
         state: 'Off'
    - condition: or
      conditions:
       - condition: template
         value_template: '{{ is_state("input_boolean.heating_timer_1_weekday", "on") and is_state("binary_sensor.workday_sensor", "on") }}'
       - condition: template
         value_template: '{{ is_state("input_boolean.heating_timer_1_weekend", "on") and is_state("binary_sensor.workday_sensor", "off") }}'
   action:
    - service: script.turn_on
      data_template:
        entity_id: script.heating_{{ states.input_select.heating_timer_1_profile.state }}

input_boolean:
  heating_timer_1_switch:
    name: Active
    icon: mdi:gesture-double-tap
  heating_timer_1_weekday:
    name: Weekdays
    icon: mdi:calendar-check
  heating_timer_1_weekday:
    name: Weekdays
    icon: mdi:calendar-check
  ########## Overrides ########
  summer_mode:
    name: "Summer Mode"
    icon: mdi:weather-sunny
  heating_holiday_mode:
    name: "Heating Holiday Mode"
    icon: mdi:beach

input_select:
  heating_timer_1_time:
    name: Time
    icon: mdi:clock
    options: !include clock.yaml
  heating_timer_1_profile:
    name: Heat Profile
    icon: mdi:thermometer-lines
    options: !include profiles.yaml

  heating_timer_1_profile:
    name: Heat Profile
    icon: mdi:thermometer-lines
    options: !include profiles.yaml

sensor:
 - platform: time_date
   display_options:
    - 'time'
    - 'date'

```

**Comment by b4dpxl** [comment]
I'd like to see this as a scheduler, instead of using automations. Please can it also consider #20154 though. With the current setup, using automations overrides any temperature set by enabling away_mode.

**Comment by efficiosoft** [comment]
A real scheduler would be the right way of solving those sorts of tasks. But again, it shouldn't be specific to the climate component IMHO.

What about the idea of using a calendar and just fetch the temperature to set from it? Reacting to away mode etc. would then be not difficult with just a few plain automations.

**Comment by b4dpxl** [comment]
Automations shouldn't have to know about away_mode, etc, for setting temperatures. The end climate component should be able to handle that.

**Comment by alex-torregrosa** [comment]
I'm currently working on a generic schedule component, I'll try to make a pull request next week.

**Comment by efficiosoft** [comment]
It's questionable whether away mode should be implemented in the climate component directly at all, since whether I'm away is something that could affect other domains as well, such as lights etc.

@alex-torregrosa I'm very interesting in what you're developing. Maybe you could link the PR to this issue to have us notified?

**Comment by pszafer** [comment]
I think there should be `scheduler class` and it should be base class for other schedulers.
Then we should create `generic_climate_scheduler` which is managed by HA schedule and then we will be able to implement scheduler for different platform.

Right now I'm writing bosch thermostat component and those devices has "built-in" week calendar with schedule to turn on comfort/eco program (for both climate and water_heater)
I think it would be better if in such situation HA would be only kind of frontend for "calendar" and for climate components without such function we can use generic_scheduler.

**Comment by dgomes** [comment]
A thermostat should have it's own scheduler, sure you can use automations to connect the scheduler to the thermostat, but that is just an extra burden for the end user.

**Comment by pszafer** [comment]
isn't gonna be worse if user create some schedule in thermostat mobile app and then try to schedule something different in HA scheduler and be surprised that it's going crazy?

**Comment by dgomes** [comment]
What I would expect, would be for the HA component to synchronise both.

**Comment by efficiosoft** [comment]
Syncing schedules between platforms sounds like a huge effort to implement properly with not much benefit. Each manufacturer allows a different number of setpoints per day, different time resolution, single/multi-week schedules, to name just two of the problems this will cause.

Why not just have a central schedule to maintain in HA (or a calendar, for instance)?

Another problem is how to deal with manual overrides, e.g. when a window is opened or if I want to overwrite the scheduled value for an hour or two.

**Comment by dgomes** [comment]
@efficiosoft I think that depends on the platform. But I see various smart thermostats exposing their schedule and HA platforms being able to incorporate that. 

Some others might just dumb down their thermostats due to lack of support.

**Comment by pszafer** [comment]
@efficiosoft I can see benefits for both approaches, but some thermostats has to have schedule set up if using auto mode. So in case we don't sync scheduler from manufacter we can't really control it (we have to ask user to use manual mode)

**Comment by efficiosoft** [comment]
@pszafer That has only usability impacts, right? So the user won't see whether schedule is active or not on the thermostat display?

Personally, I'd prefer to have HA control the thermostats, because it can make more sophisticated decisions than a thermostat can (time of day/day of week).

**Comment by b4dpxl** [comment]
> A thermostat should have it's own scheduler, sure you can use automations to connect the scheduler to the thermostat, but that is just an extra burden for the end user.

It's worth remembering that this issue was raised "specially for the generic_thermostat platform". Whatever is designed needs to support that, which has no scheduling or other capabilities beyond determining temperature and state. 

If a thermostat has its scheduler, it makes sense to me to use that instead, although I can see the benefit of being able to view/edit it from within HA. But a generic scheduler in HA could be tied into a thermostat and be used instead, assuming the thermostat's scheduler (if it exists) can be disabled.

**Comment by pszafer** [comment]
@efficiosoft Buderus setup is quite complex (I know 2 houses with similar setup).
I'm not an expert but as far as I understand buderus (IP gateway is manufactured by bosch) thermostat schedule has to be configured. In set up where there is heat pump which is used for heating (climate device) and for water (water_heater). This thermostat can have solar panel as another heating device as well. I think it is easier to create frontend and backend of manufacturer scheduler which is send to thermostat than try to implement such inteligence which knows that it controls heating device which sends water to climate and water heater.

@b4dpxl I think this is best aproach, if device has scheduler capability implement it to HA, if not use generic scheduler.

**Comment by timstanley1985** [comment]
I think you all may be over thinking this.  The scheduler should be platform agnostic.  All it needs to do is set thermostat parameters at a predetermined point in time.  This should be independent of any platform (nest etc.) configured schedule.  The user can then choose wether to use the platform and/or home assistant for scheduling

**Comment by tsvi** [comment]
Interesting how this issue pops up every time again. You want something  with less options than full fledged automation, but dynamic enough so you can change the timing easily and manually.

Definitely think this should be platform agnostic.

I imagined it as something with the following UI: a table of entries with a + sign to add more entries.
Each entry has a start time, end time (optional), service call, boolean whether it's on, input select for recurring and - sign for deletion.

**Comment by efficiosoft** [comment]
@tsvi +1 for the table approach.

However, entering a service call in every table row would add quite some overhead I think. Maybe it should be a two-step process.

1. Create some named service call templates with ``service`` and ``data``, maybe even ``data_template`` to dynamically insert the entity id
2. Select one of these templates in each of the table rows.

**Comment by balloob** [comment]
I think that a generic scheduler should work with scenes. Users can configure a scene for "active" and for "not active", or any other state that a scheduler would support.

It would be a bit more helpful if we extend our scene support to allow users to create them on the fly, including better restore support. But that's getting off topic.

**Comment by Adminiuga** [comment]
> I think you all may be over thinking this. The scheduler should be platform agnostic. All it needs to do is set thermostat parameters at a predetermined point in time. 

All scheduler should allow you to do is to call a service/script/scene. Could be `climate.set_temperature`

> This should be independent of any platform (nest etc.) configured schedule. The user can then choose wether to use the platform and/or home assistant for scheduling

I agree. Pretty much the above obsoletes any thermostat "buil-in" schedules. 

IMO if `time_pattern` trigger could be extended to become `datetime_pattern` trigger (crontab like?), that's essentially all we need for a generic scheduler? Maybe a nice UI could be a plus too :)

**Comment by b4dpxl** [comment]
It should also handle conditions. For example, I have a "schedule" which triggers at 06:30 on a workday (according to the `workday` sensor), unless a `boolean input` representing "holiday mode" is turned on.

**Comment by Francisk0** [comment]
Hi, @timstanley1985 can you upload a example with all files for begginers?

**Comment by Misiu** [comment]
> I think that a generic scheduler should work with scenes. Users can configure a scene for "active" and for "not active", or any other state that a scheduler would support.
> 
> It would be a bit more helpful if we extend our scene support to allow users to create them on the fly, including better restore support. But that's getting off-topic.

I think the solution proposed by @alex-torregrosa would be sufficient.
Of course, the ability to activate a scene at a specific time would be a nice addition but think of potential use cases. I think those will apply to many users.

-set target temperature to a specific thermostat or a group of thermostats
-set a mode to a specific thermostat or a group of thermostats
-turn on/off a single light or multiple lights
-open/close a roller shutter (or multiple roller shutters) or set it (them) to a specific percentage

For most of that, you don't need a scene. A scene is useful if you want to group multiple entities.
Creating a scene per entity won't be very user-friendly. If you want to have four temperatures in a thermostat per day (24h) you must create four scenes and if you want different behavior per thermostat and assuming you have six thermostats that give 24 scenes to create (correct me if I'm wrong).

There are at least a dozen topics on the forum about a schedule solution.

Can we please ask the devs for at least looking into this one more time?

**Comment by Jc2k** [comment]
In your specific example though you wouldn't need a scene for each of your thermostats, you could set each one differently within a single scene, I think.

But I think you are reading too much into "scenes" and what that means here. It doesn't have to be a heavyweight thing. This might not involve creating a new "Scene" entity at all. The scene 'engine' can apply (reproduce) a target state without having a predefined scene entity, allowing it to do all of your use cases. It's an implementation detail that the user might not even realise if we get it right. It's a way of turning imperative model (making a number of individual `set_X` service calls) into a declarative (describe the desired state) one, which will lend itself better to building a UI and validation.

AIUI there is already a scheduler of sorts built in, though I haven't used it (in the UI, under the Automations tab you can pick time as a trigger). Ultimately I'd like to see a nicer UI built on top of this. You would be able to use a widget to pick the time (its not clear what format to use etc) and then select one or more entities, then there would be UI to select which attributes like on/off, color, target temperature, mode, etc to change. This would respect any min/max ranges on the entitiy attributes of course. This would use the scene engine, but implicitly - it wouldn't turn it into a multi-stage process for the user.

**Comment by Misiu** [comment]
@Jc2k you are right, the implementation is transparent to the end-user, he doesn't have to know (and most of the time doesn't want to know) what is happening under the hood.
@alex-torregrosa made a [component](https://github.com/home-assistant/home-assistant/pull/22037) and a [UI panel](https://github.com/home-assistant/home-assistant-polymer/pull/2929) I think that both should be considered as a start point.

I'm aware that I could use automations, but for me, this creates more problems than solutions. For example, if I create an automation that sets the target temperature to 21 at 06:PM, but my system will be offline (power off) from 05:59 PM to 06:02 PM. To handle that case I must create another automation that will fire at the system start, checks if the current time is after that trigger time and set the target temperature.

I hope that a proper scheduler component will solve this problem by triggering at a specific hour, but it also should check every scheduler at the system start and fire it if the conditions are met.

**Comment by Jc2k** [comment]
Again though we are talking about the features as the devs see them vs as the user sees them. I’d argue that such a capability to run timers that were missed during downtime should be built into the timer automation and the scheduler is just a UI for leveraging that. Rather than adding an entirely different new set of APIs.

**Comment by Misiu** [comment]
@Jc2k I'm a .NET developer, so I sometimes write from a developer point of view.
But try to be an end-user: You set a schedule to change the temperature in your house at specific hours, 21 Celcius degrees at 06:00 PM and 18 Celcius at 09:00 PM. So if for some reason there was downtime at 06:00 PM you want that temperature to be set to 21 as soon as the system starts.
Of course, you could create a schedule every 5 minutes between 06:00 and 09:00 PM to set the temperature to 21 (just to be sure), but that's not ideal from the end-user point of view. You want the temperature to be set to 21 and that's the problem the schedule should solve.

I'm aware that this will probably require some extra API's, but I think the devs should be aware of such needs and at least should consider supporting such scenarios.

**Comment by Jc2k** [comment]
Last reply was from phone so did a crap job of explaining myself.

Let me start by saying I don't disagree with the user experiences you want at all. I want them myself 😁 I think that's important to point out because I don't want to argue or make you think I don't agree with the use cases you mention or the user experience you want to have. If anything, all I want to do is try and help build consensus or at the very least help understand where we are and why.

Also I'm not a core dev, I maintain one small integration. So you don't need to convince me of anything, and doing so won't help with your goals to get thus resolved 😄  Earlier I said you might be reading too much into the talk about scenes, but of course i might be reading too little as well (but [this](https://github.com/home-assistant/home-assistant/issues/26813) makes me thing i'm not too wrong... maybe 🥈 )

As a dev i'm sure you'll already appreciate this point but... one of the big sticking points of Home Assistant is that due to its size when you implement something and get it wrong then it's hard to change because you have 100's of users that are sensivity to changes to it. That's why the team cannot sometimes do the "obvious" and "easy" thing first and plan to do it right later. They don't want 100's of users using a feature that has any sort of planned obsolescence - it will lead to hundreds of support tickets. So big new features like this are going to be slow. We need to take the time to try and get it right, pluse everyone is busy. 

As you saw yourself a few hours ago, the generic scheduler PR does not actually support all the entities that the existing scene engine does. It doesn't support covers for example. So even if that PR had been merged it would not have handled at least one of your use cases. If the "rules" in that system were scene fragments that were passed to the scene engine just like in the PR i linked to then it would - and it would automatically get support for new entities and entity attributes that the scene engine got support for. There is even potential for sharing UI components around configuring the desired entity state with the main scene engine as well.

(To be clear i'm not saying that you as a user would even know that we were using the scene backend to do this. It would be seamless)

The alternative is maintaining that code in both components, doubling the work for an already stretched team. Again, for avoidance of doubtt. From a user POV, both of these acheive exactly the same thing. They may have the same UI. But by not using the scene engine you potentially causes duplication and/or feature disparity.

That's why i said to not read too much into the terminology. You might be *implying* a user experience was not being implied at all.

Your example of a current limitation of the time automation is a good one and I think the time automation trigger should gain [anacron](https://en.wikipedia.org/wiki/Anacron) like powers. I'd be interested to know what other people think of that idea.

**Comment by Santobert** [comment]
@Jc2k :100: % agreed!It would be great to have something like a scheduling engine based on reproducing states. This could be done using automations and scenes, but with an improved ui. Something like a subset of automations that are not visible in the automation editor, but in a schedule editor. This schedule editor could look similar to [the one of nodered](https://flows.nodered.org/node/node-red-contrib-light-scheduler).

**Comment by Misiu** [comment]
@Santobert as @Jc2k wrote the implementation is transparent to the end-user. I don't know the guts of Home Assistant, so can't say what would be the best way to build this,
[27298](https://github.com/home-assistant/home-assistant/pull/27298) got merged, [25681](https://github.com/home-assistant/home-assistant/issues/25681) is almost finished, so I guess that APIs needed for this will be almost ready.

I like the idea of anacron (didn't hear that phrase before 😉), it would be a lifesaver.
Do You think this should be a separate arch issue?

@balloob any comments? What do You think about both ideas?

**Comment by balloob** [comment]
I think that my earlier comment is still spot on: https://github.com/home-assistant/architecture/issues/148#issuecomment-502945601

This should work only with scenes, as scenes delegate the actual reproducing states to the integrations. So a scheduler would only have to deal with what scene to activate when.  

The better scene support has been added by @Santobert + others in 101 👍

**Comment by Misiu** [comment]
> This should work only with scenes, as scenes delegate the actual reproducing states to the integrations. So a scheduler would only have to deal with what scene to activate when.
> 
> The better scene support has been added by @Santobert + others in 101 👍

@balloob does this mean that the user will have to create scenes manually?
Ideally, the user should be able to use the scheduler to set temperature for a given thermostat, the UI should allow him to enter the temperature or the color of the lights or turn a switch on.
From the end-user perspective, it doesn't matter how this will be implemented, but please don't require us to manually create scenes.

Below type of UI would be a good start (taken from https://github.com/home-assistant/home-assistant-polymer/pull/2929):
![image](https://user-images.githubusercontent.com/1741838/67667003-3eb03e80-f96d-11e9-93fe-51f273eae921.png)

What I especially like about this UI is the ability to support two hours (`from` and `to`). This way the scheduler would work like [anacron ](https://en.wikipedia.org/wiki/Anacron)- if the HA would be offline at the exact time when the schedule should trigger, then after start if would check what schedule should be active during specific hours and start it.

**Comment by b4dpxl** [comment]
@Misiu that UI would be neat, but we need to have support for conditions. That’s what separates HA from a dumb thermostat or timer. For example, I have a “schedule” which only fires early on a weekday if it’s not a public holiday and I’ve not manually toggled a Boolean value indicating we’re on holiday. I also have non-heating schedules which trigger at set times if the outside temperature is above a certain threshold, or the sun hasn’t set yet. Calling a script _might_ be an option for some cases, but the basics should be handled from the UI

**Comment by Misiu** [comment]
@b4dpxl I can't disagree. Conditions would be an awesome addition, but that's an advanced usage.
In the UI I attached you can see a checkbox that probably is used to enable/disable a Scheduler. Is such advanced scenarios automation could be used to enable/disable a scheduler.

For standard scenarios, the end-user should be able to edit the scheduler, for example, set the hours during which the scheduler will be active, or set the target temperature.

Without easy to use UI this component won't be very useful - I don't think my parents (60 years +) or my wife will be happy if she will have to manually edit yaml file just to change the target temperature.

**Comment by b4dpxl** [comment]
> Without easy to use UI this component won't be very useful - I don't think my parents (60 years +) or my wife will be happy if she will have to manually edit yaml file just to change the target temperature.

Agreed @Misiu , I currently use number inputs to set my various temperatures, with a “config” Lovelace panel. Basic conditions should be possible though, as they are through the automation editor. Or at least allow advanced schedules to be created using the same scheduling engine via yaml directly, as with automations today.

**Comment by balloob** [comment]
@Misiu see https://github.com/home-assistant/home-assistant/issues/27023

At the end of the day, this is going to just be automations right? They have time triggers, conditions (including for weekday) etc. The only difference is what to do when a user changes devices on their own. For thermostats, if you change the schedule there are override options to say that the change is permanent, until the next schedule change or for next X time.

**Comment by pszafer** [comment]
I think scenes+automations are almost perfect for scheduler already.
I'd add scene with service eg `planned_temperature` and that way the real thermostat would follow schedule, and HA would just map schedule of thermostat to HA and allow to edit times, but not invoke real service like `set_temperature` or `set_mode`.

**Comment by Misiu** [comment]
> Agreed @Misiu , I currently use number inputs to set my various temperatures, with a “config” Lovelace panel. Basic conditions should be possible though, as they are through the automation editor. Or at least allow advanced schedules to be created using the same scheduling engine via yaml directly, as with automations today.

I know that this is a bit off-topic, but could You share your setup? I have 5 generic thermostats and I'm looking for a way to easily set the temperature based on a basic scheduler (until a first-class scheduler is added into HA). Thanks!

**Comment by b4dpxl** [comment]
@Misiu I have 5 [virtual_thermostats](https://github.com/b4dpxl/virtual_thermostat) (my own fork of generic_thermostat) linked into a [climate_group](https://github.com/daenny/climate_group). These are connected to powered radiator valves and ds18b20 temperature sensors. There are 2 automations to turn on (after 3 minutes due to the time to open a valve) and off the boiler itself (using a Shelly 1 connected in place of the thermostat) when any of the radiators are open.

I have a number of input_number variables for different room temperatures at different times, an input_boolean to specify when we’re on holiday (and get a lie-in), and a workday binary sensor. I also have some window sensors which set the virtual_thermostat in that room to “away” when opened; the missus likes to open windows when the heating is on :)

My automations look something like this:

- 6:30 - if a workday and not “holiday”, set the morning temperatures
- 8:00 - as above, if _not_ a workday, or on holiday
- 8:30 to 16:30 - if my daughter is in school, set her thermostat to “away”. Unset when she leaves school
10:00 - set daytime temperatures 
10:00 to 17:00 - if the house is empty, lower the temperature, and raise when someone returns 
17:00 - set the evening temperature 
20:00 - drop the current temperatures by 2 degrees
22:00 - set the night time temperature

I run my home office at slightly different temperatures/times and don’t actively heat over the weekend either. I have an RF switch by my desk that I can hit for a boost in that room, if I’m working late. 

I can’t post screenshots or get my config at the moment, but I’ll put some up later when I can.

**Comment by Misiu** [comment]
@b4dpxl can You send them to me via email or via PM on HA community? I don't want to spam this issue.

**Comment by Adminiuga** [comment]
> They have time triggers, conditions (including for weekday

Is there a reason why `time` or `time_pattern` don't include weekday?

**Comment by ciB89** [comment]
Is there a plan to introduce a UI for managing scenes? Ofc all of this is already possible with the current methods, but implementing/chaning/editing a full-scaled schedule for multiple thermostats/rooms is just a pain in the a** doing it all in yaml

**Comment by bachya** [comment]
@ciB89 As you probably saw, 0.102.0 introduced a Scene Editor. 👍 

I'm feeling the itch to start this, so I'm going to work on a back-end proof of concept. I think @balloob's idea – that a generic scheduler should focus on activating scenes – is a good MVP; we can always add other functionality (like activating automations, etc.) later on.

Here's what I'm thinking for back-end MVP:

* `scheduler` integration (we can always change the name)
* A `schedule` has three key properties: a scene ID, a start datetime, and an optional end datetime (not worrying about repeating schedules/etc. for now).
* The scene gets activated on the start datetime and, optionally, restored at the end datetime if it exists.
* Websocket endpoints for the front-end to talk to
* Services to create/delete schedules (so that automations can interact with the schedule)

Thoughts?

**Comment by Santobert** [comment]
@bachya that sounds great. I think this would be a good starting point. I'm not completely sure how to add things like conditions. Thinking about climates: when someone is at home and it's not weekend activate scene "heating" otherwise activate scene "heating_away". 

I'm now asking myself what's the difference between schedules and automations 🙈

**Comment by bachya** [comment]
>I'm now asking myself what's the difference between schedules and automations 🙈

Me too! 😆I think as we progress, it'll become clearer.

Conditions are absolutely valid and needed – I'll think on that after I have an MVP working.

**Comment by Santobert** [comment]
@bachya obviously. Let's get the first prototype running and see what's missing. I'll try to help as soon as I'm done with some other projects 😉

**Comment by Santobert** [comment]
I thought about this... 

We already know that we want to apply scenes as soon as the scheduler triggers. That's fine, so far. Let's talk about conditions. We can build these scenes depending on conditions.

For example: The climate `climate.my_climate` should change it's preset_mode according to a binary_sensor `binary_sensor.someone_at_home`. The resulting scene would look as follows:
```yaml
name: my_climate_scheduler
entities:
  climate.my_climate:
    state: heat
    preset_mode: {%- if is_state('binary_sensor.someone_at_home', 'false') %} away {% else %} none {% endif -%}
``` 

I'm using Yaml and Jinja here because it's an easy way to express my thoughts. Of course we should build this in Python. I haven't finished thinking about it yet, but that would be a useful feature and it's something automation can't do right now (`scene.create` doesn't understand templates). Just an idea...

**Comment by bachya** [comment]
Started a draft: https://github.com/home-assistant/home-assistant/pull/28938 – nowhere near close to ready for anything, but has websocket endpoints to create/update/delete/clear scheduler entries. No feedback needed yet; just an FYI.

**Comment by bachya** [comment]
Alrighty, I've got some more meat on the bone with home-assistant/home-assistant#28938. Lots of edge cases I'm missing, I'm sure, but let's see! After successfully authenticating to the websocket API, you can perform these operations.

## List Schedules

This endpoint lists all saved schedules.

Payload:
```
{
  "id": 123,
  "type": "scheduler/items"
}
```

Response:
```
{
  "id": 123,
  "type": "result",
  "success": true,
  "result": {
    "b88154e83f984f16bad3041a1bb069c4": {
      "schedule_id": "b88154e83f984f16bad3041a1bb069c4",
      "activation_context_id": "6f9991fea5c24d6f9fd5489ed104c808",
      "entity_id": "scene.turn_on_bedroom_light_to_purple",
      "start_datetime": "2020-01-01T00:00:00",
      "end_datetime": null
    },
    "86000ebf4e7d486aa94f314e483bcd42": {
      "schedule_id": "86000ebf4e7d486aa94f314e483bcd42",
      "activation_context_id": "3956cfbe5d884010afd0cf0a6bd1de2e",
      "entity_id": "scene.goodnight",
      "start_datetime": "2020-01-01T00:00:00",
      "end_datetime": "2020-01-01T12:00:00"
    }
  }
}
```

## Create Schedule

This endpoint creates a new schedule. The payload must include:

* `entity_id`: a scene entity ID
* `start_datetime`: the date/time for the scene to activate

...and can optionally include:

* `end_datetime`: the date/time for the scene to "deactivate" (i.e., for all of the entities affected by the scene to return to their previous state)

Payload:
```
{
  "id": 124,
  "type": "scheduler/items/create",
  "entity_id": "scene.turn_on_bedroom_light_to_purple",
  "start_datetime": "2020-01-01 00:00:00"
}

```

Response:
```
{
    "id": 124,
    "type": "result",
    "success": true,
    "result": {
        "schedule_id": "b88154e83f984f16bad3041a1bb069c4",
        "activation_context_id": "6f9991fea5c24d6f9fd5489ed104c808",
        "entity_id": "scene.turn_on_bedroom_light_to_purple",
        "start_datetime": "2020-01-01T00:00:00",
        "end_datetime": null
    }
}
```

## Update Schedule

This endpoint updates an existing schedule. The payload must include:

* `schedule_id`: the ID of the schedule to update

...and must include at least one of:

* `entity_id`: a scene entity ID
* `start_datetime`: the date/time for the scene to activate
* `end_datetime`: the date/time for the scene to "deactivate"

Payload:
```
{
  "id": 125,
  "type": "scheduler/items/update",
  "schedule_id": "b88154e83f984f16bad3041a1bb069c4",
  "end_datetime": "2020-01-01 12:00:00"
}

```

Response:
```
{
    "id": 125,
    "type": "result",
    "success": true,
    "result": {
        "schedule_id": "b88154e83f984f16bad3041a1bb069c4",
        "activation_context_id": "6f9991fea5c24d6f9fd5489ed104c808",
        "entity_id": "scene.turn_on_bedroom_light_to_purple",
        "start_datetime": "2020-01-01T00:00:00",
        "end_datetime": "2020-01-01T12:00:00"
    }
}
```

## Delete Schedule

This endpoint deletes an existing schedule. The payload must include:

* `schedule_id`: the ID of the schedule to update

Payload:
```
{
  "id": 126,
  "type": "scheduler/items/delete",
  "schedule_id": "b88154e83f984f16bad3041a1bb069c4",
}

```

Response:
```
{
    "id": 126,
    "type": "result",
    "success": true,
    "result": {
        "schedule_id": "b88154e83f984f16bad3041a1bb069c4",
        "activation_context_id": "6f9991fea5c24d6f9fd5489ed104c808",
        "entity_id": "scene.turn_on_bedroom_light_to_purple",
        "start_datetime": "2020-01-01T00:00:00",
        "end_datetime": "2020-01-01T12:00:00"
    }
}
```

## Clear Expired

This endpoint clears any expired schedules:

Payload:
```
{
  "id": 127,
  "type": "scheduler/items/clear_expired",
}

```

Response:
```
{
  "id": 127,
  "type": "result",
  "success": true,
  "result": {}
}
```

## Other Items to Tinker On

Some other things I'm thinking about (in no particular order):

* It's easy to create a `create_schedule` service, but need to think on whether any of these other operations are useful as services.
* Right now, if an expired schedule exists in the JSON at HASS launch, an error will be logged. That shouldn't happen (since schedules get removed from the JSON when the complete), but need to think on that further.
* I need to complete tests, linting, and typing.

**Comment by Santobert** [comment]
Woho 🎉 great work. Just to understand you right: these are one-time schedules? It would be great if we could schedule things on a regular basis, i.e. weekly. Use case: turn on my climate every workday at 5pm. Is that currently possible?

**Comment by bachya** [comment]
From [above](#issuecomment-557170118):

>* A schedule has three key properties: a scene ID, a start datetime, and an optional end datetime (not worrying about repeating schedules/etc. for now).

I definitely realize that repeating schedules are crucial to the eventual first release; wanting to get the MVP principles correct before I embark on that.

**Comment by Santobert** [comment]
Ohh sorry. I was too excited to read 🙈

**Comment by bachya** [comment]
No need to apologize! I'm excited, too. 👍

Going to get some sleep; hopefully I will dream of an elegant way to represent schedule recurrences. 😴

**Comment by Misiu** [comment]
@bachya that already looks awesome 🎉
Please also consider anacron times (as I described here https://github.com/home-assistant/architecture/issues/148#issuecomment-546864328).
If you schedule a thermostat to set the temperature to 21 degrees as 5:00 PM but you lost the power or Raspberry rebooted at 4:59 PM and was offline at 5:00 PM then the temperature won't change.
If you set the temperature to be 21 degrees between 5 and 9 PM then when the system starts it should check if there is a scheduler that had the start time before the system started and should activate it.
I think this would be super useful.

**Comment by bachya** [comment]
@Misiu There are two separate (and equally valid) items to discuss in your suggestion: (1) how to ensure schedules are validated and (2) what format recurrence rules should take.

(1) is fairly straightforward: whenever HASS starts up, it should examine existing schedules and re-establish any active/valid ones. I need to test more, but in principle, that should already be in place.

(2) deserves some discussion. Using the cron format has some advantages – its "easy" to express complex schedules – but it’s more geared toward entry in a text file. We won’t be doing that here (not primarily, at least): we’ll be using some sort of UI to generate schedules and it may force the front-end to do an excess amount of translation work if cron is used.

**Comment by bachya** [comment]
After chatting with @balloob about (2), we’re going to go with representing recurrences via RFC5545 (https://tools.ietf.org/html/rfc5545), which is an already-established standard for representing this data structure.

This will require some rework of what I had previously, but it should make for more robust MVP.

More to come.

**Comment by Santobert** [comment]
I agree with @bachya. Cron it used to run a short running task multiple times per day/week/month. It would need some modifications to express time spans with it. A native calendar format should be more natural to use. Furthermore it would be easier to build a gui around it.

**Comment by bottiger** [comment]
@bachya Without knowing the internals of home assistant. Couldn't you (re)use triggers from automations for schedules, instead of time slots? This would make them much more flexibel. 

Regarding, what to do at HA startup when using triggers. It is obviously not something that is easy to implement, by I think a good compromise would just be to implement an optional "coldStartup"/"boot" condition which the user could provide. If it evaluates to true the schedule is regarded as active on startup. 

If you only start and end schedules with timeslots you can of course search for the correct state, but you will run into problems with this approach as soon as someone wants something more advanced. Like, don't turn the heat on when I'm not home.

Personally I like the trigger approach because it should be familiar to the user (and developers!). And if the user does not provide any "coldStartup" condition the schedular will just adjust itself within a cycle.

**Comment by bottiger** [comment]
@bachya One more thing. I took a look at you initial implementation. It seems like at the end of a schedule go back to the state before the scheduler started. While it's a good initial approach this can be a problem is there are more than two states of the entity.

I know that this probably isn't something one should focus on initially, but it's worth keeping in mind

**Comment by efficiosoft** [comment]
> If you only start and end schedules with timeslots you can of course search for the correct state, but you will run into problems with this approach as soon as someone wants something more advanced. Like, don't turn the heat on when I'm not home.

The correct concept for making schedules depend on entity states would be conditions, not triggers. They can be checked anytime, both on startup and when one of the participating entities changes its state. At least this is something I learned during the development of Schedy and what users have to understand first.

**Comment by bottiger** [comment]
@efficiosoft Really? The main use case I have in my head is that I want my lights/heating/etc to be on during the evening and turn off during the night.

* I want it to turn on at a specific time, or when I arrive home early.
* I want it to turn off at 22:00, or when I push my "goodnight button" to shut off everything

If I represent "I'm home" as a condition I guess the system would need to poll for that all the time, in contrast to if you define it as a trigger. If it's a trigger it will be "triggered" once whenever the state of "I'm home" state changes.

And I'm not sure how you represent "I have pushed the goodnight button" as a condition either. I guess you could set a boolean value which then resets during the night, but it feels like hack.

**Comment by efficiosoft** [comment]
> If I represent "I'm home" as a condition I guess the system would need to poll for that all the time, in contrast to if you define it as a trigger. If it's a trigger it will be "triggered" once whenever the state of "I'm home" state changes.

The condition would only need to be reevaluated when the home state changes as well. The issue with using triggers is that their purpose is to react to events that happen at some point (I'm not talking about events as on the HA event bus). However, a schedule needs to be able to determine whether it should be active at any time, not just be activated in response to some event. Otherwise, it wouldn't be able to be activated at startup and we could just use automations as we have them today.

**Comment by bachya** [comment]
For MVP, the scheduler is going to focus on scheduling scenes (activating them and optionally unwinding them).

Automations and scripts aren’t a great fit:

1. Automations and scripts can be infinitely complex; if we want to “unwind” an automation, HASS would have to understand all possible unwinding scenarios, which is a huge undertaking.
2. Automations aren’t great candidates to be scheduled because, as already pointed out, their triggers and conditions already form a sort of “schedule.“ Want to turn on your lights when you get home? Create an automation.

I see a world where people can use all of these simultaneously. In my case, my house has:

1. scenes that will benefit from the scheduler (allowing me to remove extraneous automations that exist just to trigger/unwind those scenes)
2. automations that don’t need to be “scheduled” and can be left as-is
3. Schedy instances to handle the hyper-complex cases not covered by (1) and (2)

I can see a v2 of the scheduler that includes support for conditions (in which I will definitely attempt to make use of the existing condition support for automations), but that will be after the MVP.

I’m closing in on a v1 of recurrences (allowing for more complicated repetitions of schedules). Will hopefully share today.

**Comment by efficiosoft** [comment]
> I can see a v2 of the scheduler that includes support for conditions (in which I will definitely attempt to make use of the existing condition support for automations), but that will be after the MVP.

With that in place, the schedules could have functionality similar to that of Schedy, but with the known semantics of HA conditions. Can't wait to see something like that.

**Comment by efficiosoft** [comment]
@bachya Just something that came to my mind: Isn't solely relying on conditions, even for start and end time and repititions an option as well? A quick glance at [the docs](https://www.home-assistant.io/docs/scripts/conditions/#time-condition) show that it could benefit from some more options, e.g. for month or year, but actually the time condition is something users are already familiar with and that way, a schedule could be activated by anything we have a condition for (esp. cool for sun or state) without any extra work or new syntax.

**Comment by bottiger** [comment]
I'm going to try to recap how I understand this:

@efficiosoft As I understand it, the main property you would like a system to have is: Be able to have a deterministic starts & stops so you can determine if it should be running without knowing the current state. i.e. after reboot/at startup. I think this is a reasonable property to pursue, albeit I don't feel I need it.

The property I am pursuing is instead: I really would like to transition my devices to the state they should be in now. The typical example is when you want to enable scene A for 5 minutes. You store the current scene B, transition to scene A, and after 5 minutes you can go back to scene B. However, depending on the circumstances this may or may not be correct.

In other words: I would like to write an automation which says: "Turn the light green for 5 minutes, and the go back to whatever state the scheduler thinks you should be in". Today the automation needs to keep track of every possible state the light could be in after 5 minutes have past.

@bachya Regarding "unwinding", I don't understand why it's so complex. Obviously you cannot unwind to something you don't know, but you can unwind to an active schedule - which is exactly the property I am looking for.

Admittedly there are some limitations, for instance: What happens if there are multiple active schedules ? (Why I would like to give them an optional priority)

> Want to turn on your lights when you get home? Create an automation.

I think the statement above is a bit misunderstood. I can of course create an automation which turns on the correct scene when I come home, but if I ever change that scene I may have trouble coming back (unless the scheduler has started in the meantime). That said, I really think this is a minor issue for now.

**Comment by efficiosoft** [comment]
> I'm going to try to recap how I understand this:

Thanks, that's great since I sometimes am not sure whether I can clearly express what I want to in English.

> @efficiosoft As I understand it, the main property you would like a system to have is: Be able to have a deterministic starts & stops so you can determine if it should be running without knowing the current state. i.e. after reboot/at startup. I think this is a reasonable property to pursue, albeit I don't feel I need it.

Exactly, I think it's crucial for a schedule to be applicable solely based on current time/state/whatever and not on a trigger.

> The property I am pursuing is instead: I really would like to transition my devices to the state they should be in now. The typical example is when you want to enable scene A for 5 minutes. You store the current scene B, transition to scene A, and after 5 minutes you can go back to scene B. However, depending on the circumstances this may or may not be correct.
>
> In other words: I would like to write an automation which says: "Turn the light green for 5 minutes, and the go back to whatever state the scheduler thinks you should be in". Today the automation needs to keep track of every possible state the light could be in after 5 minutes have past.

That's an important feature imho, however I would say it's not strictly related to schedules, it's rather a manual intervention. The reverting of a scene activation could get its own service call (maybe `scene.reset` or something) which would then execute the unwinding logic @bachya implemented. Your automation could then activate a scene, delay for five minutes and finally call the reset service. It then wouldn't matter whether the previous state was achieved by a schedule or not.

Does this make sense?

**Comment by bachya** [comment]
@efficiosoft:

>@bachya Just something that came to my mind: Isn't solely relying on conditions, even for start and end time and repititions an option as well? A quick glance at the docs of the time condition show that it could benefit from some more options, e.g. for month or year, but actually the time condition is something users are already familiar with and that way, a schedule could be activated by anything we have a condition for (esp. cool for sun or state) without any extra work or new syntax.

The only challenge is that this syntax doesn't handle recurrences (9-10am every other Thursday). Additionally, we're envisioning some sort of calendar UI for this (using [RFC5545](https://tools.ietf.org/html/rfc5545) as the standard), so schedules won't be created through YAML. That doesn't mean we can't look at using the concept for dynamic times – like sunset – post-MVP.

**Comment by bachya** [comment]
Alrighty, round 2 has just been pushed to [the PR](https://github.com/home-assistant/home-assistant/pull/28938) – this is my first attempt at recurrences (using [RFC5545](https://tools.ietf.org/html/rfc5545) as the standard).

After successfully authenticating to the websocket API, you can perform these operations.

## List Schedules

This endpoint lists all saved schedules.

Payload:
```
{
  "id": 123,
  "type": "scheduler/schedules"
}
```

Response:
```
{
  "id": 123,
  "type": "result",
  "success": true,
  "result": {
    "b88154e83f984f16bad3041a1bb069c4": {
      "schedule_id": "b88154e83f984f16bad3041a1bb069c4",
      "activation_context_id": "6f9991fea5c24d6f9fd5489ed104c808",
      "entity_id": "scene.turn_on_bedroom_light_to_purple",
      "start_datetime": "2020-01-01T00:00:00",
      "duration": null,
      "rrule": null
    },
    "86000ebf4e7d486aa94f314e483bcd42": {
      "schedule_id": "86000ebf4e7d486aa94f314e483bcd42",
      "activation_context_id": "3956cfbe5d884010afd0cf0a6bd1de2e",
      "entity_id": "scene.goodnight",
      "start_datetime": "2020-01-01T00:00:00",
      "duration": null,
      "rrule": null
    }
  }
}
```

## Create Schedule

This endpoint creates a new schedule. The payload must include:

* `entity_id`: a scene entity ID
* `start_datetime`: the date/time for the scene to activate

...and can optionally include:

* `duration`: the length of time (in seconds) that the schedule lasts for. Including it will reverse the scene activation when that duration is hit; omitting it will leave reversal out.
* `rrule`: an [RFC5545](https://tools.ietf.org/html/rfc5545)-compliant recurrence string

Let's look at some examples.

### Turn on a scene this Christmas at noon

Payload:
```
{
  "id": 124,
  "type": "scheduler/schedules/create",
  "entity_id": "scene.turn_on_bedroom_light_to_purple",
  "start_datetime": "2019-12-25 12:00:00"
}
```

Response:

```
{
    "id": 124,
    "type": "result",
    "success": true,
    "result": {
        "schedule_id": "b88154e83f984f16bad3041a1bb069c4",
        "activation_context_id": "6f9991fea5c24d6f9fd5489ed104c808",
        "entity_id": "scene.turn_on_bedroom_light_to_purple",
        "start_datetime": "2019-12-25 12:00:00"
        "duration": null,
        "rrule": null
    }
}
```

### Turn on a scene this Christmas at noon and reverse/undo it 1 hour later

Payload:
```
{
  "id": 124,
  "type": "scheduler/schedules/create",
  "entity_id": "scene.turn_on_bedroom_light_to_purple",
  "start_datetime": "2019-12-25 12:00:00"
  "duration": 3600,
}
```

Response:

```
{
    "id": 124,
    "type": "result",
    "success": true,
    "result": {
        "schedule_id": "b88154e83f984f16bad3041a1bb069c4",
        "activation_context_id": "6f9991fea5c24d6f9fd5489ed104c808",
        "entity_id": "scene.turn_on_bedroom_light_to_purple",
        "start_datetime": "2019-12-25 12:00:00"
        "duration": 3600,
        "rrule": null
    }
}
```

### Turn on a scene this Christmas at noon, reverse/undo it 1 hour later, and repeat that every 2 days

Payload:
```
{
  "id": 124,
  "type": "scheduler/schedules/create",
  "entity_id": "scene.turn_on_bedroom_light_to_purple",
  "start_datetime": "2019-12-25 12:00:00"
  "duration": 3600,
  "rrule": "FREQ=DAILY;INTERVAL=2"
}
```

Response:

```
{
    "id": 124,
    "type": "result",
    "success": true,
    "result": {
        "schedule_id": "b88154e83f984f16bad3041a1bb069c4",
        "activation_context_id": "6f9991fea5c24d6f9fd5489ed104c808",
        "entity_id": "scene.turn_on_bedroom_light_to_purple",
        "start_datetime": "2019-12-25 12:00:00"
        "duration": 3600,
        "rrule": "FREQ=DAILY;INTERVAL=2"
    }
}
```

Note that the `rrule` format is defined by RFC5545; in our case, we'd anticipate that it would be created by some sort of calendar UI.

## Update Schedule

This endpoint updates an existing schedule. The payload must include:

* `schedule_id`: the ID of the schedule to update

...and must include at least one of:

* `entity_id`: a scene entity ID
* `start_datetime`: the date/time for the scene to activate
* `duration`: the length of time (in seconds) that the schedule lasts for. Including it will reverse the scene activation when that duration is hit; omitting it will leave reversal out.
* `rrule`: an RFC5545-compliant recurrence string

Payload:
```
{
  "id": 125,
  "type": "scheduler/schedules/update",
  "schedule_id": "b88154e83f984f16bad3041a1bb069c4",
  "duration": 7200
}

```

Response:
```
{
    "id": 125,
    "type": "result",
    "success": true,
    "result": {
        "schedule_id": "b88154e83f984f16bad3041a1bb069c4",
        "activation_context_id": "6f9991fea5c24d6f9fd5489ed104c808",
        "entity_id": "scene.turn_on_bedroom_light_to_purple",
        "start_datetime": "2019-12-25 12:00:00"
        "duration": 7200,
        "rrule": "FREQ=DAILY;INTERVAL=2"
    }
}
```

## Delete Schedule

This endpoint deletes an existing schedule. The payload must include:

* `schedule_id`: the ID of the schedule to update

Payload:
```
{
  "id": 126,
  "type": "scheduler/schedules/delete",
  "schedule_id": "b88154e83f984f16bad3041a1bb069c4",
}

```

Response:
```
{
    "id": 126,
    "type": "result",
    "success": true,
    "result": {
        "schedule_id": "b88154e83f984f16bad3041a1bb069c4",
        "activation_context_id": "6f9991fea5c24d6f9fd5489ed104c808",
        "entity_id": "scene.turn_on_bedroom_light_to_purple",
        "start_datetime": "2020-01-01T00:00:00",
        "duration": 7200,
        "rrule": "FREQ=DAILY;INTERVAL=2"
    }
}
```

**Comment by Santobert** [comment]
That looks really great! Thanks @bachya

**Comment by Misiu** [comment]
> * Still thinking on what to do if HASS reboots in the middle of a schedule

And what in situation when HASS reboots just before schedule should start and starts after schedule should start (power off at 4:58 PM, start at 5:10 PM, the schedule should start at 5:00 PM)?

**Comment by bottiger** [comment]
@efficiosoft is of the (reasonable) oppinion that you should be able to test if the schedule should be running at any time. i.e. at 5:10 PM - when HASS has started - it can test all the defined schedules so see if they should be running or not.

**Comment by efficiosoft** [comment]
> The only challenge is that this syntax doesn't handle recurrences (9-10am every other Thursday). Additionally, we're envisioning some sort of calendar UI for this (using RFC5545 as the standard), so schedules won't be created through YAML. That doesn't mean we can't look at using the concept for dynamic times – like sunset – post-MVP.

I see that the RFC5545 syntax is easier to parse for using it in a GUI. It seems like the general direction of Hass goes away from YAML and towards JSON and GUI-only management (which I personally don't like at all), but that's another story.

**Comment by electrofloat** [comment]
> It seems like the general direction of Hass goes away from YAML and towards JSON and GUI-only management (which I personally don't like at all), but that's another story.

The exact opposite: https://www.home-assistant.io/blog/2019/10/10/release-100/#goodbye-json-

**Comment by efficiosoft** [comment]
> The exact opposite: https://www.home-assistant.io/blog/2019/10/10/release-100/#goodbye-json-

Ah, good to know. But I was not talking about the particular serialization format used to enter data in the GUI, rather about the general fact that there are things you can only manage using the GUI (like these schedules or the ZWave config imported from YAML into JSON storage).

**Comment by bachya** [comment]
@efficiosoft Appreciate the feedback. We'll keep an eye on things; post-MVP, if there's a need for some other entry method (services, a YAML file, etc.), we'll certainly take a look.

**Comment by balloob** [comment]
As we did with the person integration etc, we can allow schedules to be defined in YAML.

**Comment by Ambidexter211** [comment]
Keep in mind that an overall schedule feature would be able to work with any integration. Turn on / dim / off my lights at certain times, and when certain events happen. Turn up the volume on my media player when there are 10 people shopping in my store, etc.

But then we end up with regular automations, more or less.

I suppose it would be useful if automations could be...nested or atomic, that is, separate pieces of automations that are either subroutines/functions or macros. Once I get a function that turns on a lamp and sets the brightness to 70% brightness (or +20% brightness), I could call that atomic routine from any automation by passing the light or other entity, and the brightness value, and get a return of the current brightness of the entity.

This is certainly easy to do in Python, but maybe we need a library of such Python primitives that are callable by YAML automations?

Anyway, thanks for all your work folks!

Cheers,

Ambi

**Comment by yabuts** [comment]
> Interesting how this issue pops up every time again. You want something with less options than full fledged automation, but dynamic enough so you can change the timing easily and manually.
> 
> Definitely think this should be platform agnostic.
> 
> I imagined it as something with the following UI: a table of entries with a + sign to add more entries.
> Each entry has a start time, end time (optional), service call, boolean whether it's on, input select for recurring and - sign for deletion.

Tzi I whould love to have a ui scheduler and also be able to integrate with your shabbos yomtov issue maluch etc,

**Comment by Misiu** [comment]
@bachya awesome piece of work! what's the status of this PR? Any plans to move this forward? No pressure at all, just wondering when this might be added. I think that many users will be happy to have Scheduler in Home Assistant 🙂

**Comment by bachya** [comment]
@Misiu It will be a bit; appreciate your patience. @balloob wanted to review the use case architecture before we proceeded further and he has a lot on his plate.

**Comment by antonverburg** [comment]
Maybe I miss something, but did somebody already think about the fact that some processes take a lot of time? For example, if heating a room, a user preferably wants to tell the system when the room has to be on temperature. This means that the scheduler has to start long time before this time with heating the room, so that it is on temperature in time. I actually wrote a script that reads a google-calendar via the calendar integration, and that can predict how long to heat before it is on temperature (depending on outer temperature, and in future I also want to add wind speed & direction). Did we already take this prediction-features in account for the design of the scheduler?

**Comment by bachya** [comment]
@antonverburg Prediction is way outside of an MVP. On day 1, I would expect people to use the scheduler with these predictions implied (e.g., if you know that it takes your house an hour to heat up to where you want it by 5pm, add a schedule that starts at 4pm).

**Comment by Misiu** [comment]
@bachya what's left to do. The API looks almost complete.
With backend ready work on front end could be started.

**Comment by bachya** [comment]
@Misiu I've paused on this after discussion with @balloob – it needs more consideration before we can say the back- end is “done,”

**Comment by fjufju** [comment]
@bachya any updates/decisions about this? I really hope scheduler will be added in near future releases. This would be an awesome feature!

**Comment by bachya** [comment]
Given the pause on this, I’ve moved on to other projects.

**Comment by fjufju** [comment]
thank you for info.
I hope this will be added soon :)

**Comment by Misiu** [comment]
There is an alternative for now:
https://github.com/nielsfaber/scheduler-component
https://github.com/nielsfaber/scheduler-card

Detailed description [here](https://community.home-assistant.io/t/scheduler-card-custom-component/217458)

**Comment by virgilm** [comment]
This is absolutely needed in HA! How can you have a home automation system without a scheduler?

Options I found:
- Scheduler component above: nice, actively developed, but still not fully functional for my needs.
- Schedy - really nice and flexible, but no front end. Also, appdaemon, not HA.

**Comment by marcelveldt** [comment]
Something like this perhaps ?
https://github.com/marcelveldt/homeassistant-scheduler/tree/master/custom_components/schedules

- A schedule has a before and after property which can be filled with either time (e.g. 22:00:00) of a sun notation (sunrise + 01:00:00)
- A schedule also has a condition, which accepts templating. For example the state of the workday sensor must be true
- You can add/update/remove schedules through service calls.
- Schedules will be represented as binary sensor, so you can easily check for state in every script/automation etc.
- It would be pretty easy to add a frontend for this too.
- It would be pretty easy to create device automations based on the schedules so you'll have a generic solution for all hass components.

The good thing about it is that it would require zero core changes as this is done with building blocks that were already there.

Please have a look and test it if you like.
If there's enough interest I can submit a PR to core.

**Comment by nielsfaber** [comment]
@marcelveldt I am working on a similar project.
There is a thread about it [here](https://community.home-assistant.io/t/scheduler-card-custom-component/217458).
I would be happy to collaborate on this, to prevent making double efforts..
My project consists of a card+custom_component combination..
The component has quite some overlap with yours, but has some differences as well (mostly provisions to store multiple timepoints + actions per entity).
Then again i like your coding style much better (i think my component is a bit messy written, but i am also not really a python hero).
May take a few steps to get there, but would be great if we could merge things to a solid piece of work!

**Comment by marcelveldt** [comment]
@nielsfaber sounds good but my time is a bit sparse atm. In a few weeks I'll be able to help out.

**Comment by nielsfaber** [comment]
@marcelveldt Any updates on this?
I still would like to join forces.
Could you [PM](https://community.home-assistant.io/u/neliss) me for discussing next steps?

**Comment by GamerClassN7** [comment]
Any updaes reguarding the implementation of climate schedukling ? this is really one of missing core feature :(

**Comment by frenck** [comment]
This architecture issue is old, stale, and possibly obsolete. Things changed a lot over the years. Additionally, we have been moving to discussions for these architectural discussions.

For that reason, I'm going to close this issue.

../Frenck
